runtime/compiler/control/HookedByTheJit.cpp

```diff
@@ -2343,8 +2336,78 @@ static bool methodsAreRedefinedInPlace()
    return false;
    }
 
-// Hack markers
-#define VM_PASSES_SAME_CLASS_TWICE 1
+namespace { // file-local
+
+/**
+ * \brief A pair of corresponding classes related by redefinition,
+ * distinguished from each other in two ways: old/new and fresh/stale.
+ *
+ * NOTE: Neither set of terminology, i.e. neither old/new nor fresh/stale as
+ * used here, is necessarily consistent with terminology elsewhere in the VM.
+ */
+struct ElaboratedClassPair
+   {
+   /**
+    * \brief The original class that existed before HCR.
+    *
+    * This is always different from newClass.
+    */
+   TR_OpaqueClassBlock *oldClass;
+
+   /**
+    * \brief The class that was created in order to carry out HCR.
+    *
+    * This is always different from oldClass.
+    */
+   TR_OpaqueClassBlock *newClass;
+
+   /**
+    * \brief The class whose methods have the old bytecode.
+    *
+    * This is always different from freshClass. It is equal to newClass if the
+    * class has been redefined in-place, and oldClass otherwise.
+    */
+   TR_OpaqueClassBlock *staleClass;
+
+   /**
+    * \brief The class whose methods have the new bytecode.
+    *
+    * This is always different from staleClass. It is equal to oldClass if the
+    * class has been redefined in-place, and newClass otherwise.
+    */
+   TR_OpaqueClassBlock *freshClass;
+   };
+
+void setElaboratedClassPair(ElaboratedClassPair *ecp, J9JITRedefinedClass *classPair)
+   {
+   // The VM tells us the old J9Class and the fresh one (which here it calls "new").
+   J9Class *oldJ9Class = classPair->oldClass;
+   J9Class *freshJ9Class = classPair->newClass;
+   J9Class *staleJ9Class = freshJ9Class->replacedClass;
+
+   ecp->oldClass = TR::Compiler->cls.convertClassPtrToClassOffset(oldJ9Class);
+   ecp->freshClass = TR::Compiler->cls.convertClassPtrToClassOffset(freshJ9Class);
+   ecp->staleClass = TR::Compiler->cls.convertClassPtrToClassOffset(staleJ9Class);
+
+   TR_ASSERT_FATAL(
+      ecp->freshClass != ecp->staleClass,
+      "fresh and stale classes are the same: %p",
+      ecp->freshClass);
+
+   TR_ASSERT_FATAL(
+      ecp->oldClass == ecp->freshClass || ecp->oldClass == ecp->staleClass,
+      "oldClass %p matches neither freshClass %p nor staleClass %p",
+      ecp->oldClass,
+      ecp->freshClass,
+      ecp->staleClass);
+
+   // Don't try to predict whether classes should be redefined in-place.
+   // Instead just check whether this one was in fact redefined in-place.
+   ecp->newClass =
+      ecp->oldClass == ecp->freshClass ? ecp->staleClass : ecp->freshClass;
```

Comment:
Personally I tend to think that adding unnecessary parentheses is how we end up making things harder to read without them. It might even misleadingly suggest to a reader that the precedence order is the opposite of what it really is. I also think it's interesting that I haven't been asked to parenthesize the entire ternary expression to clarify its precedence relative to the assignment, which in my mind isn't really any more or less readily understood than equality vs ternary. (Even though the precedence *rule* for ternary/assignment is definitely less obvious: they're at the same precedence level, and it works the way we expect because they're right-associative)

That said, this kind of thing is always a judgment call, and I doubt that this comment will do much to convince anyone to judge this situation differently than they already would have. At the very least I hope I haven't convinced you to ask for more unnecessary parentheses as opposed to fewer :sweat_smile:

In any case, it's not a huge deal, so I will add an extra pair (or even two) if you insist

FWIW the effort of writing this comment and the size of it both seem excessive to me compared to just adding a pair of parentheses, but I just felt it was worth saying this once :shrug:

Reply:
> Even though the precedence rule for ternary/assignment is definitely less obvious: they're at the same precedence level

I would have to agree there - my immediate intuition had been that the conditional ternary had higher precedence than assignment.  I'm happy to defer to your judgment and will withdraw this suggestion.